### PR TITLE
Sorted tokens

### PR DESCRIPTION
**Pull request by wolverineks**
sorted-tokens

[YOMO-317](https://emurgo.atlassian.net/jira/software/c/projects/YOMO/boards/5?modal=detail&selectedIssue=YOMO-317)

[YOMO-317]: https://emurgo.atlassian.net/browse/YOMO-317?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/yoroi-wallets/cardano/api.ts [hunk 1]
```diff
@@ -184,3 +175,86 @@ export const fetchCurrentPrice = async (currency: CurrencySymbol, config: Backen
...
+  return {
+    id: toAssetFingerprint(token.metadata.policyId, token.metadata.assetName),
+    group: token.metadata.policyId,
+    name: hexToAscii(token.metadata.assetName),
+    decimals: token.metadata.numberOfDecimals,
+    description: token.metadata.longName ?? '',
+  }
+}
+
+import AssetFingerprint from '@emurgo/cip14-js'
```
Comment (banklesss) [inline src/yoroi-wallets/cardano/api.ts]: Why here?

Comment (wolverineks) [reply src/yoroi-wallets/cardano/api.ts]: copy pasta

File: src/Send/AssetSelectorScreen/AssetSelectorScreen.tsx [hunk 2]
```diff
@@ -28,18 +29,19 @@ export const AssetSelectorScreen = () => {
   const [matcher, setMatcher] = React.useState('')
   const navigation = useNavigation<TxHistoryRouteNavigation>()
   const {tokenSelected, allTokensSelected} = useSend()
+  const tokenInfos = useTokenInfos({
+    wallet,
+    tokenIds: Amounts.toArray(balances).map(({tokenId}) => tokenId),
+  })
+  const sortedTokenInfos = tokenInfos
+    .sort(descending((tokenInfo) => tokenInfo.name.toLocaleLowerCase()))
```
Comment (wolverineks) [inline src/Send/AssetSelectorScreen/AssetSelectorScreen.tsx]: figure this out

File: src/Send/AssetSelectorScreen/AssetSelectorScreen.tsx [hunk 3]
```diff
@@ -94,40 +95,45 @@ export const AssetSelectorScreen = () => {
...
           </Text>
+
           <Text
             numberOfLines={1}
             ellipsizeMode="middle"
             style={{color: COLORS.TEXT_INPUT}}
             testID="tokenFingerprintText"
           >
-            {tokenInfo.isDefault ? '' : getTokenFingerprint(tokenInfo)}
+            {isPrimary ? '' : tokenInfo.id}
```
Comment (wolverineks) [inline src/Send/AssetSelectorScreen/AssetSelectorScreen.tsx]: tokenInfo.fingerprint

File: src/Send/AssetSelectorScreen/AssetSelectorScreen.tsx [hunk 4]
```diff
@@ -152,14 +158,12 @@ const SearchInput = (props) => {
...
-  normalize(token.metadata.longName ?? '').includes(matcher) ||
-  normalize(token.identifier).includes(matcher) ||
-  normalize(token.metadata.assetName).includes(matcher) ||
-  normalize(token.metadata.policyId).includes(matcher)
+const matches = (tokenInfo: TokenInfo, matcher: string) =>
+  normalize(tokenInfo.ticker ?? '').includes(matcher) ||
+  normalize(tokenInfo.name).includes(matcher) ||
+  normalize(tokenInfo.id).includes(matcher) ||
+  normalize(tokenInfo.description ?? '').includes(matcher) ||
+  normalize(tokenInfo.group ?? '').includes(matcher)
```
Comment (wolverineks) [inline src/Send/AssetSelectorScreen/AssetSelectorScreen.tsx]: tokenInfo.fingerprint

File: src/Send/ConfirmScreen/ConfirmScreen.tsx [hunk 5]
```diff
@@ -182,9 +182,11 @@ const TokenTotals = ({yoroiUnsignedTx}: {yoroiUnsignedTx: YoroiUnsignedTx}) => {
 
 const Amount = ({amount}: {amount: YoroiAmount}) => {
   const wallet = useSelectedWallet()
-  const tokenInfo = useTokenInfo({wallet, tokenId: amount.tokenId})
+  const tokenId = amount.tokenId
+  if (tokenId.includes('.')) throw new Error(`invalid tokenId: ${tokenId}`)
```
Comment (wolverineks) [inline src/Send/ConfirmScreen/ConfirmScreen.tsx]: remove this

File: src/Send/Context/SendContext.tsx [hunk 6]
```diff
@@ -37,10 +37,13 @@ export const SendProvider = ({
   const actions = React.useRef<SendActions>({
     receiverChanged: (receiver) => dispatch({type: 'receiverChanged', receiver}),
     amountChanged: (amount) => dispatch({type: 'amountChanged', amount}),
-    tokenSelected: (tokenId) => dispatch({type: 'tokenSelected', tokenId}),
+    tokenSelected: (tokenId) => {
+      if (tokenId.includes('.')) throw new Error(`invalid tokenId: ${tokenId}`)
```
Comment (wolverineks) [inline src/Send/Context/SendContext.tsx]: remove this

File: src/Send/SendScreen/SendAllWarning.tsx [hunk 7]
```diff
@@ -20,10 +20,12 @@ export const SendAllWarning = ({
 }: SendAllWarningProps) => {
   const strings = useStrings()
   const wallet = useSelectedWallet()
-  const tokenInfo = useTokenInfo({wallet, tokenId: selectedTokenIdentifier})
-  const assetNameOrId = truncateWithEllipsis(getAssetDenominationOrId(tokenInfo), 20)
+  if (selectedTokenIdentifier !== wallet.primaryTokenInfo.id && !selectedTokenIdentifier.includes('.'))
```
Comment (wolverineks) [inline src/Send/SendScreen/SendAllWarning.tsx]: remove this

File: src/TxHistory/AssetList/AssetList.tsx [hunk 8]
```diff
@@ -49,48 +53,52 @@ export const AssetList = (props: Props) => {
...
   onPress?: () => void
 }
 
-const AssetItem = ({amount, onPress}: AssetItemProps) => {
+const AssetItem = ({tokenInfo, onPress}: AssetItemProps) => {
   const wallet = useSelectedWallet()
-  const tokenInfo = useTokenInfo({wallet, tokenId: amount.tokenId})
+  const tokenId = tokenInfo.id
+  // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
+  if (tokenId.includes('.')) throw new Error(`invalid tokenId: ${tokenId}`)
```
Comment (wolverineks) [inline src/TxHistory/AssetList/AssetList.tsx]: remove this

File: src/TxHistory/AssetList/AssetList.tsx [hunk 9]
```diff
@@ -49,48 +53,52 @@ export const AssetList = (props: Props) => {
...
 
       <Middle>
         <Text numberOfLines={1} ellipsizeMode="middle" style={styles.tokenInfo} testID="tokenInfoText">
-          {getAssetDenominationOrId(tokenInfo)}
+          {tokenInfo.ticker ?? tokenInfo.name}
         </Text>
 
         <Text numberOfLines={1} ellipsizeMode="middle" style={styles.tokenName} testID="tokenFingerprintText">
-          {tokenInfo.isDefault ? '' : getTokenFingerprint(tokenInfo)}
+          {isPrimary ? '' : tokenInfo.id}
```
Comment (wolverineks) [inline src/TxHistory/AssetList/AssetList.tsx]: fingerprint

File: src/yoroi-wallets/cardano/api.ts [hunk 10]
```diff
@@ -40,10 +42,21 @@ export const fetchNewTxHistory = async (
     request.addresses.length <= config.TX_HISTORY_MAX_ADDRESSES,
     'fetchNewTxHistory: too many addresses',
   )
-  const response = (await fetchDefault('v2/txs/history', request, config)) as Array<RawTransaction>
+  const transactions = (await fetchDefault('v2/txs/history', request, config)) as Array<RawTransaction>
+
+  // validate tokenId
+  transactions.forEach(({inputs, outputs}) =>
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/api.ts]: remove this

File: src/TxHistory/TxDetails/AssetList.tsx [hunk 11]
```diff
@@ -54,23 +54,26 @@ const AssetRow = ({styles, entry, backColor, onSelect}: AssetRowProps) => {
...
         <Text style={styles.assetName}>
-          {tokenInfo.isDefault
-            ? getTicker(tokenInfo)
-            : getName(tokenInfo) ?? intl.formatMessage(messages.unknownAssetName)}
+          {isPrimary ? tokenInfo.ticker : tokenInfo.name ?? intl.formatMessage(messages.unknownAssetName)}
         </Text>
 
         <Text style={styles.assetMeta} ellipsizeMode="middle" numberOfLines={1}>
-          {tokenInfo.isDefault ? '' : getTokenFingerprint(tokenInfo)}
+          {isPrimary ? '' : tokenInfo.id}
```
Comment (wolverineks) [inline src/TxHistory/TxDetails/AssetList.tsx]: fingerprint

File: src/TxHistory/TxDetails/AssetList.tsx [hunk 12]
```diff
@@ -54,23 +54,26 @@ const AssetRow = ({styles, entry, backColor, onSelect}: AssetRowProps) => {
   const intl = useIntl()
   const wallet = useSelectedWallet()
   const tokenInfo = useTokenInfo({wallet, tokenId: entry.identifier})
+  const tokenId = tokenInfo.id
+  // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
+  if (tokenId.includes('.')) throw new Error(`invalid tokenId: ${tokenId}`)
```
Comment (wolverineks) [inline src/TxHistory/TxDetails/AssetList.tsx]: remove this

File: src/yoroi-wallets/types/tokens.ts [hunk 13]
```diff
@@ -67,22 +67,28 @@ export type DefaultAsset = Token & {
...
-  policyId: string
-  longName?: string
-  ticker?: string
-}
-
-// https://github.com/cardano-foundation/cardano-token-registry#semantic-content-of-registry-entries
-export type TokenRegistryEntry = {
-  subject: string
-  name: string
+  id: string // asset fingerprint
```
Comment (wolverineks) [inline src/yoroi-wallets/types/tokens.ts]: not fingerprint

File: src/Send/AssetSelectorScreen/AssetSelectorScreen.tsx [hunk 14]
```diff
@@ -152,14 +157,8 @@ const SearchInput = (props) => {
...
-const matches = (token: Token, matcher: string) =>
-  normalize(decodeHexAscii(token.metadata.assetName) ?? '').includes(matcher) ||
-  normalize(getTokenFingerprint(token) ?? '').includes(matcher) ||
-  normalize(token.metadata.ticker ?? '').includes(matcher) ||
-  normalize(token.metadata.longName ?? '').includes(matcher) ||
-  normalize(token.identifier).includes(matcher) ||
-  normalize(token.metadata.assetName).includes(matcher) ||
-  normalize(token.metadata.policyId).includes(matcher)
+const matches = (tokenInfo: TokenInfo, matcher: string) =>
+  Object.keys(tokenInfo).map((key) => normalize(key).includes(matcher))
```
Comment (michalszorad) [inline src/Send/AssetSelectorScreen/AssetSelectorScreen.tsx]: I think you meant `Object.keys(tokenInfo).some`

File: src/Send/SendScreen/SendScreen.tsx [hunk 15]
```diff
@@ -94,7 +94,7 @@ export const SendScreen = () => {
       addressInput: receiver,
       amount,
       sendAll,
-      selectedTokenInfo: tokenInfo,
+      selectedTokenInfo: token,
```
Comment (michalszorad) [inline src/Send/SendScreen/SendScreen.tsx]: What about `selectedToken: token`?

File: src/TxHistory/AssetList/AssetList.tsx [hunk 16]
```diff
@@ -29,18 +28,23 @@ export const AssetList = (props: Props) => {
...
   const handleSearch = () => Alert.alert(strings.soon, strings.soon)
 
+  const tokenInfos = useTokenInfos({
+    wallet,
+    tokenIds: Amounts.toArray(balances).map(({tokenId}) => tokenId),
+  })
+
+  const sortedTokenInfos = tokenInfos
+    .sort(alpha((tokenInfo) => tokenInfo.name.toLocaleLowerCase()))
+    .sort(startsWith((tokenInfo) => tokenInfo.id === wallet.primaryTokenInfo.id))
```
Comment (michalszorad) [inline src/TxHistory/AssetList/AssetList.tsx]: 👍

File: src/Send/AssetSelectorScreen/AssetSelectorScreen.tsx (hunk 14 above)
Comment (wolverineks) [reply src/Send/AssetSelectorScreen/AssetSelectorScreen.tsx]: for simple fixes you can use the suggestion tool, or just make the change

File: src/yoroi-wallets/cardano/ShelleyWallet.ts [hunk 17]
```diff
@@ -1292,3 +1302,20 @@ const keys: Array<keyof WalletJSON> = [
   'isEasyConfirmationEnabled',
   'lastGeneratedAddressIndex',
 ]
+
+export const primaryTokenInfo = {
```
Comment (michalszorad) [inline src/yoroi-wallets/cardano/ShelleyWallet.ts:1306]: `export const primaryTokenInfo: Record<'mainnet' | 'testnet', TokenInfo > = {...}`

File: src/yoroi-wallets/types/tokens.ts [hunk 18]
```diff
@@ -67,22 +67,27 @@ export type DefaultAsset = Token & {
...
+
+export type LegacyToken = {
+  networkId: number
+  isDefault: boolean
+  /**
+   * For Ergo, this is the tokenId (box id of first input in tx)
+   * for Cardano, this is policyId || assetName
+   * Note: we don't use null for the primary token of the chain
+   * As some blockchains have multiple primary tokens
+   */
```
Comment (stackchain) [inline src/yoroi-wallets/types/tokens.ts:90]:
```
```

File: src/yoroi-wallets/cardano/ShelleyWallet.ts [hunk 19]
```diff
@@ -1292,3 +1302,20 @@ const keys: Array<keyof WalletJSON> = [
...
+    ticker: 'ADA',
+  } as TokenInfo,
+  testnet: {
+    id: '',
+    name: 'TADA',
+    decimals: 6,
+    description: '',
+    ticker: 'TADA',
+  } as TokenInfo,
+}
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/ShelleyWallet.ts:1321]:
```suggestion
} as const
```

File: src/Send/SendScreen/SendScreen.tsx [hunk 20]
```diff
@@ -192,17 +192,16 @@ export const SendScreen = () => {
...
         </TouchableOpacity>
 
         <Checkbox
           checked={sendAll}
           onChange={sendAllChanged}
-          text={
-            tokenInfo.isDefault ? strings.checkboxSendAllAssets : strings.checkboxSendAll({assetId: assetDenomination})
-          }
+          // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
+          text={token.isDefault ? strings.checkboxSendAllAssets : strings.checkboxSendAll({assetId: assetDenomination})}
```
Comment (stackchain) [inline src/Send/SendScreen/SendScreen.tsx]: `isPrimaryToken`

**Review by stackchain — COMMENTED** [review]
⬆️

File: src/yoroi-wallets/types/tokens.ts (hunk 18 above)
Comment (stackchain) [inline src/yoroi-wallets/types/tokens.ts]:
```suggestion
```

Comment (stackchain) [reply src/yoroi-wallets/types/tokens.ts]: ~removing since it's outdated~

**Review by stackchain — APPROVED** [review]

**Pull request merged**